pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -575,14 +575,24 @@ func (p *Pod) equivalentToWorkload(wl *kueue.Workload, jobPodSets []kueue.PodSet
 		return false
 	}
 
-	for i := range wl.Spec.PodSets {
-		if i >= len(jobPodSets) {
-			return true
-		}
-		if !workloadFinished && wl.Spec.PodSets[i].Count < jobPodSets[i].Count {
-			return false
+	// Match the current state of pod sets
+	// to the pod set info in the workload
+	for i := range jobPodSets {
+		jobPodSetIsFound := false
+		for j := i; j < len(wl.Spec.PodSets); j++ {
```

Comment:
I don't think we can have a linear implementation without a separate var. 

In the proposed implementation `j` will be equal `len(wl.Spec.PodSets)-1` if a role is not found. 

I had something like this:
```go
        // Match the current state of pod sets
	// to the pod set info in the workload
	lastFoundPodSetIdx := -1
	for i := range jobPodSets {
		jobPodSetIsFound := false
		for j := lastFoundPodSetIdx + 1; j < len(wl.Spec.PodSets); j++ {
			jobPodSet := jobPodSets[i]
			wlPodSet := wl.Spec.PodSets[j]

			if wlPodSet.Name == jobPodSet.Name {
				if !workloadFinished && wlPodSet.Count < jobPodSet.Count {
					return false
				}
				jobPodSetIsFound = true
				lastFoundPodSetIdx = j
			}
		}
		// If actual pod set info has a role that workload doesn't have,
		// consider workload not an equivalent to the pod group
		if !jobPodSetIsFound {
			return false
		}
	}
```

But decided to drop it because it was harder to read and understand. If you want I can go back to this implementation.

Reply:
If a role is not found it's game over, just return false

Reply:
It's the same implementation, just without an extra variable

Reply:
Ah, I guess I could just add something like that:

if j == len(wl.Spec.PodSets) -1 { j = -1 }

Reply:
Ah, yes. You're right. Updated the implementation.